url49584c1d41_Arrem_GGS_tgz-pJ8-com_gamezgalaxy_test_console_MainJ8/src/com/gamezgalaxy/GGS/networking/packets/PacketManager.java

```diff
@@ -0,0 +1,183 @@
+import com.gamezgalaxy.GGS.networking.packets.minecraft.MOTD;
+import com.gamezgalaxy.GGS.networking.packets.minecraft.Message;
+import com.gamezgalaxy.GGS.networking.packets.minecraft.Ping;
+import com.gamezgalaxy.GGS.networking.packets.minecraft.PosUpdate;
+import com.gamezgalaxy.GGS.networking.packets.minecraft.SetBlock;
+import com.gamezgalaxy.GGS.networking.packets.minecraft.SpawnPlayer;
+import com.gamezgalaxy.GGS.networking.packets.minecraft.TP;
+import com.gamezgalaxy.GGS.networking.packets.minecraft.UpdateUser;
+import com.gamezgalaxy.GGS.networking.packets.minecraft.Welcome;
+import com.gamezgalaxy.GGS.server.Server;
+
+public class PacketManager {
+	
+	protected Packet[] packets = new Packet[] {
+		new Connect(this),
+		new DespawnPlayer(this),
+		new FinishLevelSend(this),
+		new GlobalPosUpdate(this),
+		new Kick(this),
+		new LevelSend(this),
+		new LevelStartSend(this),
+		new Message(this),
+		new MOTD(this),
+		new Ping(this),
+		new PosUpdate(this),
+		new SetBlock(this),
+		new SpawnPlayer(this),
+		new TP(this),
+		new UpdateUser(this),
+		new Welcome(this),
+		new GET(this)
+	};
+	
+	protected ServerSocket serverSocket;
+	
+	protected Thread reader;
+	
+	/**
+	 * The server this PacketManager belongs to
+	 */
+	public Server server;
+	
+	/**
+	 * The constructor for the PacketManager
+	 * @param instance
+	 *                The server that this PacketManager will belong to
+	 */
+	public PacketManager(Server instance) {
+		this.server = instance;
+		try {
+			serverSocket = new ServerSocket(this.server.Port);
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
+	}
+	
+	/**
+	 * Get a packet this PacketManager handles.
+	 * @param opCode
+	 *              The OpCode for the packet
+	 * @return
+	 *        The packet found, if no packet is found, then it will
+	 *        return null.
+	 */
+	public Packet getPacket(byte opCode) {
+		for (Packet p : packets) {
+			if (p.ID == opCode)
+				return p;
+		}
+		return null;
+	}
+	
+	/**
+	 * Get a packet this PacketManager handles.
+	 * @param name
+	 *            The name of the packet
+	 * @return
+	 *        The packet found, if no packet is found, then it will
+	 *        return null.
+	 */
+	public Packet getPacket(String name) {
+		for (Packet p : packets) {
+			if (p.name.equalsIgnoreCase(name))
+				return p;
+		}
+		return null;
+	}
+	
+	/**
+	 * Have the PacketManager start listening for clients
+	 * on the port provided by the {@link PacketManager#server}
+	 */
+	public void StartReading() {
+		reader = new Read();
+		reader.start();
+		server.Log("Listening on port " + server.Port);
+	}
+	
+	/**
+	 * Stop listening for clients.
+	 */
+	@SuppressWarnings("deprecation")
+	public void StopReading() {
+		reader.stop();
+		try {
+			serverSocket.close();
+		} catch (IOException e1) {
+			// TODO Auto-generated catch block
+			e1.printStackTrace();
+		}
+		try {
+			reader.join();
+		} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)